connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectWorkerIntegrationTest.java

```diff
@@ -195,7 +197,22 @@ public void testRestartFailedTask() throws Exception {
     public void testBrokerCoordinator() throws Exception {
         ConnectorHandle connectorHandle = RuntimeHandles.get().connectorHandle(CONNECTOR_NAME);
         workerProps.put(DistributedConfig.SCHEDULED_REBALANCE_MAX_DELAY_MS_CONFIG, String.valueOf(5000));
-        connect = connectBuilder.workerProps(workerProps).build();
+        Properties brokerProps = new Properties();
+
+        // Find a free port and use it in the Kafka broker's listeners config. We can't use port 0 in the listeners
+        // config to get a random free port because in this test we want to stop the Kafka broker and then bring it
+        // back up and listening on the same port in order to verify that the Connect cluster can re-connect to Kafka
+        // and continue functioning normally. If we were to use port 0 here, the Kafka broker would most likely listen
+        // on a different random free port the second time it is started.
```

Comment:
Sorry, my bad for not clarifying at the outset. There's a couple of reasons - one is that the previous implementation seemed fairly messy where we were checking the bound ports after broker startup and then modifying the broker config and using the static port in the listener configuration on broker restarts only if the user hadn't already configured listener configs. So, for instance, this `startOnlyKafkaOnSamePorts` would not work as expected if the user defines a SASL or SSL listener with port 0. The other reason is that the `KafkaClusterTestKit` being leveraged here would need some refactors to allow changing broker configs after it has already been instantiated - currently, it only supports customising broker configs in its builder. All in all, it just seems a lot cleaner to move the responsibility of using a fixed port in the listeners configuration to the clients of the embedded Kafka cluster in case they want to test functionality involving offline brokers.

Reply:
Fair enough, that works for me!
The one small change I'd like to see is that, in tests where we set listeners with a fixed non-zero port, we make sure to explicitly set the number of brokers to one (I know that this matches the current default, but we may change the default, and it'd be nice to make this test resilient against that).

Reply:
Actually, the more I think about it, the more brittle this seems. We're introducing a way to override the listeners property in the `KafkaClusterTestKit` class, but that only works for single-node clusters, and the use cases are fairly thin (IIUC it's not actually necessary to be able to override the `listeners` property for tests that involve changing the security protocols in use, which basically just leaves us with being able to reuse ports across restarts).

I'm still not completely against this (since if used responsibly, you can still do what you need to), but I wonder if it might be cleaner to do the refactoring in the `KafkaClusterTestKit` class instead, both to make it reusable and to allow for multi-node cases.

Reply:
> We're introducing a way to override the listeners property in the KafkaClusterTestKit class, but that only works for single-node clusters

The override ability is also currently used for setting up SASL / SSL listeners (which can work in multi-node cases using port 0). 

> IIUC it's not actually necessary to be able to override the listeners property for tests that involve changing the security protocols in use

I think you're suggesting adding an API to the `KafkaClusterTestKit.Builder` class to be able to conveniently setup SASL / SSL listeners instead of each test setting all the individual listener related configs?

Reply:
Correct me if I'm wrong, but there doesn't seem to be a case yet that requires overriding the `listeners` property that can't also be addressed by leaving that property as-is and only touching the `listener.security.protocol.map` property, except for reusing ports across restarts.

So yes, although it's possible to override `listeners` to set up SASL/SSL, this isn't required, and IMO it's actually cleaner if you do things by altering the security protocol map instead.